SABT_MainUnit/MD4.c

```diff
@@ -25,293 +36,325 @@
-      }
-      break;
-
-    case MD4_STATE_SAY_MISTAKES:
-      sprintf(bufff, "#%d", num_mistakes);
-      play_mp3(LANG_FILESET, bufff);
-      play_mp3(MODE_FILESET,MP3_MISTAKES);
-      md4_current_state = MD4_STATE_ASK_FOR_GUESS;
-      break;
-
-    case MD4_STATE_ASK_FOR_GUESS:
-      play_mp3(MODE_FILESET,MP3_GUESS);
-      md4_current_state = MD4_STATE_WAIT_INPUT;
-      break;
-
-    case MD4_STATE_WAIT_INPUT:
-      if(got_input)
-      {
-        got_input = false;
-        md4_current_state = MD4_STATE_PROC_INPUT;
-      }
-      break;
-
-    case MD4_STATE_PROC_INPUT:
-      // nothing was entered so we repeat the word
-      if (last_cell == 0)
-      {
-        md4_current_state = MD4_STATE_EVALUATE_GAME;
-      } else if (md4_valid_letter(last_cell))
-      { // set entered_letter in valid_letter(), but return true or false
-        char buff[7];
-        sprintf(buff, "%c", entered_letter);
-        play_mp3(LANG_FILESET,buff);
-        md4_current_state = MD4_STATE_CHECK_MATCH;
-      } else
-      {
-        play_mp3(LANG_FILESET,MP3_INVALID);
-        num_mistakes++;
-        md4_current_state = MD4_STATE_EVALUATE_GAME;
-      }
-      break;
 
-    case MD4_STATE_CHECK_MATCH:
-      // place_letter() returns true if the letter guessed is found in
-      // the word, false otherwise. If true, it will place the entered_letter
-      // into input_word.
-      if (place_letter())
-      {
-        play_mp3(LANG_FILESET,MP3_YES);
-      } else
-      {
-        play_mp3(LANG_FILESET, MP3_NO);
-        num_mistakes++;
-      }
-
-      md4_current_state = MD4_STATE_EVALUATE_GAME;
-      break;
+void md4_reset(void) {
+    PRINTF("*** MD4 - one player hangman ***\n\r");
+    
+    // Global variables
+    set_mode_globals(&script_english, LANG_FILESET, MODE_FILESET);
+    
+    // State variables
+    md_next_state = STATE_INTRO;
+    md_last_dot = NO_DOTS;
+    md_input_ready = false;
+    md_input_valid = false;
+    md_incorrect_tries = 0;
+}
 
-    case MD4_STATE_EVALUATE_GAME:
-      if (!strncmp(input_word, current_word, strlen(current_word)))
-      {
-        game_status = 1;
-        play_mp3(MODE_FILESET,MP3_YOU_WIN);  // "you have guessed the word!"
-      } else if (num_mistakes == 7)
-      {
-        game_status = 1;
-        play_mp3(MODE_FILESET,MP3_YOU_LOSE); // "you have made 7 mistakes the word you missed was"
-      }
+void md4_main(void) {
+	int num_words;	
+    switch (md_next_state) {
+        case STATE_INTRO:
+			shuffle(dict.num_words, dict.index_array);
+			dict.index = 0;
+            play_mp3(MODE_FILESET, MP3_INTRO);
+            md_next_state = STATE_GENQUES;
+            break;
+            
+        case STATE_GENQUES:
+			
+            num_words = dict.num_words;
+            if (dict.index >= num_words - 1){
+                shuffle(num_words, dict.index_array);
+                dict.index = 0;
+                }
+            chosen_word = dict.words[dict.index_array[dict.index]];

```

Comment:
If in three hint mode and word is three letters, the word is immediately guessed.  Would be nice to add a check here that hints are less than letter count.  Also am curious what happens with double letters (like with "book" and three letters are given, would both "o"s be given too, thus all letters being given?).